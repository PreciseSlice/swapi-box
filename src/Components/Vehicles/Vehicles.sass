.vehicles 
  color: #feda4a