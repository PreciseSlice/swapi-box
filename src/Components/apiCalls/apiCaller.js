export const callFetch = async url => {
  const getData = await fetch(url);
  const jsonData = await getData.json();

  return await jsonData;
};

export const getFilms = async url => {
  try {
    const fetchFilmData = await callFetch(url);

    const title = await fetchFilmData.title;
    const episodeId = await fetchFilmData.episode_id;
    const openingCrawl = await fetchFilmData.opening_crawl;
    const releaseDate = await fetchFilmData.release_date;

    return await { title, episodeId, openingCrawl, releaseDate };
  } catch (er) {
    const error = new Error('getFilms failed to fetch filmData');
    return error;
  }
};

export const getPeople = async url => {
  try {
    const fetchPeopleData = await callFetch(url);

    const peopleMap = fetchPeopleData.results.map(async person => {
      const name = person.name;

      const fetchHomeworldData = await callFetch(person.homeworld);
      const homeworld = fetchHomeworldData.name;
      const population = fetchHomeworldData.population;

      const fetchSpeciesData = await callFetch(person.species);
      const speciesName = fetchSpeciesData.name;

      return { name, homeworld, population, speciesName };
    });
    return Promise.all(peopleMap);
  } catch (er) {
    const error = new Error('getPeople failed to fetch peopleData');
    return error;
  }
};

export const getPlanets = async url => {
  try {
    const fetchPlanetData = await callFetch(url);

    const planetMap = fetchPlanetData.results.map(async planet => {
      const name = planet.name;
      const terrain = planet.terrain;
      const population = planet.population;
      const climate = planet.climate;

      const residentMap = planet.residents.map(async resident => {
        const fetchResidentData = await callFetch(resident);

        return fetchResidentData.name;
      });

      const mapReturn = await Promise.all(residentMap);
      const residents = mapReturn.join(', ');

      return { name, terrain, population, climate, residents };
    });
    return Promise.all(planetMap);
  } catch (er) {
    const error = new Error('getPlaenets failed to fetch plaenetData');
    return error;
  }
};

export const getVehicles = async url => {
  try {
    const fetchVehicleData = await callFetch(url);

    const vehicleMap = fetchVehicleData.results.map(async vehicle => {
      const name = vehicle.name;
      const model = vehicle.model;
      const vehicleClass = vehicle.vehicle_class;
      const passengers = vehicle.passengers;

      return { name, model, vehicleClass, passengers };
    });
    return Promise.all(vehicleMap);
  } catch (er) {
    const error = new Error('getVehicles failed to fetch vehicleData');
    return error;
  }
};
