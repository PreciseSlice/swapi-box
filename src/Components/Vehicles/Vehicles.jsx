import React, { Component } from 'react';
import './Vehicles.css';
import { getVehicles } from '../apiCalls/apiCaller';
import Card from '../Card/Card.jsx';
//import PropTypes from 'prop-types';

class Vehicles extends Component {
  async componentDidMount() {
    const { setVehicles, vehicleData } = this.props;
    console.log(vehicleData)
    if (vehicleData.length < 1) {
      const vehicleData = await getVehicles();
      setVehicles(vehicleData);
    }
  }

  render() {
    const { vehicleData } = this.props;
    
    const renderCards = vehicleData.map( vehicle => (
      <Card vehicle={vehicle} key={vehicle.name} />
    ));
      
    if (vehicleData) {
      return <div className="vehicle-container">{renderCards}</div>
    } else {
      return null;
    }
  }
};

// Vehicles.propTypes = {
//   planetData: PropTypes.arrayOf(
//     PropTypes.shape({
//       name: PropTypes.string.isRequired,
//       terrain: PropTypes.string.isRequired,
//       population: PropTypes.number.isRequired, //somtimes a string
//       climate: PropTypes.string.isRequired,
//       residents: PropTypes.arrayOf().isRequired,
//     }).isRequired
//   ).isRequired
// };

export default Vehicles;