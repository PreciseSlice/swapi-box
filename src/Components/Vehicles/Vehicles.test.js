/* eslint-disable */
import React from 'react';
import ReactDOM from 'react-dom';
import Vehicles from './Vehicles';
import { shallow, mount } from 'enzyme';
import { cleanVehicles } from '../apiCalls/apiMockData';

describe('Vehicles', () => {
  let wrapper;
  let mockFunction;
  let secondMockFunction;

  beforeEach(() => {
    mockFunction = jest.fn();
    secondMockFunction = jest.fn();
    wrapper = mount(
      <Vehicles
        setVehicles={mockFunction}
        vehicleData={cleanVehicles}
        clickHandler={secondMockFunction}
        favorites={cleanVehicles}
      />
    );
  });

  it('Exist and matches snapshot', () => {
    expect(wrapper).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });

  it('should get people data if it does not have it, by calling getPeople.', async () => {
    wrapper = shallow(
      <Vehicles
        setVehicles={mockFunction}
        vehicleData={[]}
        clickHandler={secondMockFunction}
        favorites={cleanVehicles}
      />
    );

    await wrapper.instance().componentDidMount();
    expect(mockFunction).toHaveBeenCalled();
  });

  it('renders a card for each object in the array it is passed', () => {
    expect(wrapper.find('.card').length).toEqual(10);
  });
});
