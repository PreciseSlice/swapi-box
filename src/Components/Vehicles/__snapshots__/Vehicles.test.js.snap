// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Vehicles Exist and matches snapshot 1`] = `
<Vehicles
  clickHandler={[Function]}
  favorites={
    Array [
      Object {
        "model": "Digger Crawler",
        "name": "Sand Crawler",
        "passengers": "30",
        "vehicleClass": "wheeled",
      },
      Object {
        "model": "T-16 skyhopper",
        "name": "T-16 skyhopper",
        "passengers": "1",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "X-34 landspeeder",
        "name": "X-34 landspeeder",
        "passengers": "1",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "Twin Ion Engine/Ln Starfighter",
        "name": "TIE/LN starfighter",
        "passengers": "0",
        "vehicleClass": "starfighter",
      },
      Object {
        "model": "t-47 airspeeder",
        "name": "Snowspeeder",
        "passengers": "0",
        "vehicleClass": "airspeeder",
      },
      Object {
        "model": "TIE/sa bomber",
        "name": "TIE bomber",
        "passengers": "0",
        "vehicleClass": "space/planetary bomber",
      },
      Object {
        "model": "All Terrain Armored Transport",
        "name": "AT-AT",
        "passengers": "40",
        "vehicleClass": "assault walker",
      },
      Object {
        "model": "All Terrain Scout Transport",
        "name": "AT-ST",
        "passengers": "0",
        "vehicleClass": "walker",
      },
      Object {
        "model": "Storm IV Twin-Pod",
        "name": "Storm IV Twin-Pod cloud car",
        "passengers": "0",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "Modified Luxury Sail Barge",
        "name": "Sail barge",
        "passengers": "500",
        "vehicleClass": "sail barge",
      },
    ]
  }
  setVehicles={[Function]}
  vehicleData={
    Array [
      Object {
        "model": "Digger Crawler",
        "name": "Sand Crawler",
        "passengers": "30",
        "vehicleClass": "wheeled",
      },
      Object {
        "model": "T-16 skyhopper",
        "name": "T-16 skyhopper",
        "passengers": "1",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "X-34 landspeeder",
        "name": "X-34 landspeeder",
        "passengers": "1",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "Twin Ion Engine/Ln Starfighter",
        "name": "TIE/LN starfighter",
        "passengers": "0",
        "vehicleClass": "starfighter",
      },
      Object {
        "model": "t-47 airspeeder",
        "name": "Snowspeeder",
        "passengers": "0",
        "vehicleClass": "airspeeder",
      },
      Object {
        "model": "TIE/sa bomber",
        "name": "TIE bomber",
        "passengers": "0",
        "vehicleClass": "space/planetary bomber",
      },
      Object {
        "model": "All Terrain Armored Transport",
        "name": "AT-AT",
        "passengers": "40",
        "vehicleClass": "assault walker",
      },
      Object {
        "model": "All Terrain Scout Transport",
        "name": "AT-ST",
        "passengers": "0",
        "vehicleClass": "walker",
      },
      Object {
        "model": "Storm IV Twin-Pod",
        "name": "Storm IV Twin-Pod cloud car",
        "passengers": "0",
        "vehicleClass": "repulsorcraft",
      },
      Object {
        "model": "Modified Luxury Sail Barge",
        "name": "Sail barge",
        "passengers": "500",
        "vehicleClass": "sail barge",
      },
    ]
  }
>
  <div
    className="vehicle-container"
  >
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "Digger Crawler",
          "name": "Sand Crawler",
          "passengers": "30",
          "vehicleClass": "wheeled",
        }
      }
      highlighted="highlighted"
      key="Sand Crawler"
    >
      <div
        className="card highlighted"
      >
        <h1>
          Sand Crawler
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: Digger Crawler
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 30
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: wheeled
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "T-16 skyhopper",
          "name": "T-16 skyhopper",
          "passengers": "1",
          "vehicleClass": "repulsorcraft",
        }
      }
      highlighted="highlighted"
      key="T-16 skyhopper"
    >
      <div
        className="card highlighted"
      >
        <h1>
          T-16 skyhopper
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: T-16 skyhopper
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 1
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: repulsorcraft
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "X-34 landspeeder",
          "name": "X-34 landspeeder",
          "passengers": "1",
          "vehicleClass": "repulsorcraft",
        }
      }
      highlighted="highlighted"
      key="X-34 landspeeder"
    >
      <div
        className="card highlighted"
      >
        <h1>
          X-34 landspeeder
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: X-34 landspeeder
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 1
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: repulsorcraft
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "Twin Ion Engine/Ln Starfighter",
          "name": "TIE/LN starfighter",
          "passengers": "0",
          "vehicleClass": "starfighter",
        }
      }
      highlighted="highlighted"
      key="TIE/LN starfighter"
    >
      <div
        className="card highlighted"
      >
        <h1>
          TIE/LN starfighter
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: Twin Ion Engine/Ln Starfighter
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 0
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: starfighter
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "t-47 airspeeder",
          "name": "Snowspeeder",
          "passengers": "0",
          "vehicleClass": "airspeeder",
        }
      }
      highlighted="highlighted"
      key="Snowspeeder"
    >
      <div
        className="card highlighted"
      >
        <h1>
          Snowspeeder
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: t-47 airspeeder
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 0
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: airspeeder
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "TIE/sa bomber",
          "name": "TIE bomber",
          "passengers": "0",
          "vehicleClass": "space/planetary bomber",
        }
      }
      highlighted="highlighted"
      key="TIE bomber"
    >
      <div
        className="card highlighted"
      >
        <h1>
          TIE bomber
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: TIE/sa bomber
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 0
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: space/planetary bomber
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "All Terrain Armored Transport",
          "name": "AT-AT",
          "passengers": "40",
          "vehicleClass": "assault walker",
        }
      }
      highlighted="highlighted"
      key="AT-AT"
    >
      <div
        className="card highlighted"
      >
        <h1>
          AT-AT
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: All Terrain Armored Transport
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 40
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: assault walker
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "All Terrain Scout Transport",
          "name": "AT-ST",
          "passengers": "0",
          "vehicleClass": "walker",
        }
      }
      highlighted="highlighted"
      key="AT-ST"
    >
      <div
        className="card highlighted"
      >
        <h1>
          AT-ST
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: All Terrain Scout Transport
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 0
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: walker
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "Storm IV Twin-Pod",
          "name": "Storm IV Twin-Pod cloud car",
          "passengers": "0",
          "vehicleClass": "repulsorcraft",
        }
      }
      highlighted="highlighted"
      key="Storm IV Twin-Pod cloud car"
    >
      <div
        className="card highlighted"
      >
        <h1>
          Storm IV Twin-Pod cloud car
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: Storm IV Twin-Pod
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 0
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: repulsorcraft
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
    <Card
      clickHandler={[Function]}
      data={
        Object {
          "model": "Modified Luxury Sail Barge",
          "name": "Sail barge",
          "passengers": "500",
          "vehicleClass": "sail barge",
        }
      }
      highlighted="highlighted"
      key="Sail barge"
    >
      <div
        className="card highlighted"
      >
        <h1>
          Sail barge
        </h1>
        <ul>
          <li
            key="function now() { [native code] }0"
          >
            model: Modified Luxury Sail Barge
          </li>
          <li
            key="function now() { [native code] }1"
          >
            passengers: 500
          </li>
          <li
            key="function now() { [native code] }2"
          >
            vehicleClass: sail barge
          </li>
        </ul>
        <button
          onClick={[Function]}
        >
          favorite
        </button>
      </div>
    </Card>
  </div>
</Vehicles>
`;
